lib/src/conflicts.rs

```diff
@@ -218,6 +220,19 @@ pub fn materialize_merge_result(
     single_hunk: Merge<ContentHunk>,
     output: &mut dyn Write,
 ) -> std::io::Result<()> {
+    // The following code assumes that each version of the file contains 0 or
+    // more lines, each ending with `\n`. For now, we force this to be true.
+    // TODO(#3795): A better solution for this.
+    let single_hunk: Merge<ContentHunk> = single_hunk
+        .into_iter()
+        .map(|ContentHunk(mut side)| {
+            if !side.is_empty() && !side.ends_with(b"\n") {
+                side.push(b'\n');
+            };
+            ContentHunk(side)
+        })
+        .collect();
```

Comment:
Can't we do padding when writing hunks or added/removed contents? It will help implement "a better solution" later.

It might render eol-only change as "different", but I think that's correct.

Reply:
What do you mean by padding?

Reply:
something like
```rust
write_all(data);
if !data.ends_with(b'\n') {
    // TODO: better solution ...
    write_all(b"\n");
}
```

Reply:
First of all, what we are trying to avoid is:

    <<<<<<< Conflict 1 of 1
    %%%%%%% Changes from base to side #1
    -base+++++++ Contents of side #2
    right>>>>>>> Conflict 1 of 1 ends

The most naive interpretation of your suggestion (putting the padding at the end of the function) wouldn't help with that AFAICT.

------------------

What makes more sense (but I'm leaning against from) is to put the padding inside the conflict processing, around https://github.com/martinvonz/jj/pull/3977/files#diff-15b48dd183d7f89959f15b80cee2dcae0bb7099e1e934497f74c73384315fa6bR254 (this is what I tried doing first, the new line is a remnant of this). Note that each side of the hunk must be padded, and this will only matter if there is a conflict hunk at the very end of the file.

I think that approach is doable, and it has the advantage that, if there is no conflict at the end of the file, the file would not be mangled.

However, the reason I moved it to the beginning is that my plan was to replace padding with the more complicated transformation I described in #3795. That transformation would need to be applied both inside and outside conflicts, so the beginning of the function seems to be the right place for it.

--------

Is moving the padding inside conflict processing what you meant?

Reply:
> Is moving the padding inside conflict processing what you meant?

Yes. If I understand the problem, the logic is correct to the point of `let merge_result = files::merge(&slices)`, and the rendering code is broken. We'll need to fix `output.write_all(&left.0)?;`, `write_diff_hunks()`, etc. to be compatible with line-oriented conflict format, by inserting `"\n"` (or `"\n\\No newline at end of file`?)

Reply:
I'm a little hesitant to move in this direction, since it doesn't immediately square with my ideas for #3795.

Since the rendering code is internal, I was thinking that we can simply declare that there is a precondition for it to work correctly: the input has to be 0 or more newline-terminated lines.

In my mind, this is similar to the fact that our rendering code is broken (produces results that cannot be parsed correctly) when the input contains conflict markers.

I'm now thinking of just preparing the full solution from #3795, since it treats both issues. I'm worried that you'll oppose it since it's similar in spirit to this PR: each side is encoded to satisfy the 2 conditions (newlines & not having and conflict markers) before the file conflict is split into hunks. Then, when the conflicted file is read, the sides would be decoded after parsing the hunks (that part is entirely missing from this PR).

It currently seems difficult to me to move the encoding part of the full solution to the rendering code, but perhaps it'll be clearer when I write it up.

Let me know if you have any suggestions; otherwise we can continue the discussion when I make the new PR.

Reply:
> Since the rendering code is internal, I was thinking that we can simply declare that there is a precondition for it to work correctly: the input has to be 0 or more newline-terminated lines.
> 
> In my mind, this is similar to the fact that our rendering code is broken (produces results that cannot be parsed correctly) when the input contains conflict markers.

To me, the rendering code is broken, so we should fix it rather than introducing lossy transformation. Perhaps, nested conflict markers can be escaped somehow (e.g. by using hunk syntax?)

> I'm now thinking of just preparing the full solution from #3795, since it treats both issues.

nit: wrong issue number?

> I'm worried that you'll oppose it since it's similar in spirit to this PR: each side is encoded to satisfy the 2 conditions (newlines & not having and conflict markers) before the file conflict is split into hunks.

I'm not sure why you want to preprocess the raw inputs. It might be slightly easier, but is wrong in that `"foo"` and `"foo\n"` no longer produce a conflict? That might be okay in practice, but would be weird if conflicted file had no conflicts inside.

If I remember correctly, one of the proposed solutions was to insert something like `\ No newline at end of file` and decode it by `parse_conflict()`. I think `parse_conflict_hunk()` will roughly do:
1. split by conflict markers
2. parse contents and hunks (where "No EOL" will be processed)
3. collect them into `Merge<ContentHunk>`

And it makes sense that `materialize_merge_result()` does the reverse.

Reply:
Once that is done, my preliminary plan for #3223 is to materialize deleted sides as `\JJ There is no file at this path on this side of the conflict. It may have been deleted.`

Reply:
> is wrong in that "foo" and "foo\n" no longer produce a conflict? That might be okay in practice, but would be weird if conflicted file had no conflicts inside.

By the way, this is a good point, I hadn't initially realized that there could be conflicts that are no longer conflicts after adding the newline at the end, potentially resulting in occasional problems like #3223.